### PR TITLE
Ignore certain objects at numpydoc validation

### PR DESCRIPTION
**Pull request by jalalium**
This PR aims to ignore the following objects from numpydoc validation:

- If class is `SemanticParser` and `name.startswith('_visit_')`
- If class is `SyntacticParser` and `name.startswith('_visit_')`
- if file in `codegen.printing` and `name.startswith('_print_')`

Additionally, the function `should_ignore` can be used should one need to extend the objects to ignore.

### CODE REVIEW

**Review by EmilyBourne — APPROVED** [review]

**Comment by EmilyBourne** [comment]
@jalalium Is this tested?

**Comment by jalalium** [comment]
> @jalalium Is this tested?

Yes, as you can see from [f7a2f36](https://github.com/pyccel/pyccel/pull/1325/commits/f7a2f361318bcbaf463680b82f9f0f01ac7c42b2), the validator ignores the objects you mentioned

File: ci_tools/process_results.py [hunk 1]
```diff
@@ -10,25 +10,38 @@
                         help='Github step summary')
 args = parser.parse_args()
 
+all_codes = ['GL01', 'GL02', 'GL03', 'GL05', 'GL06', 'GL07', 'GL08', 'GL09',
+             'GL10', 'SS01', 'SS02', 'SS03', 'SS04', 'SS05', 'SS06', 'ES01',
+             'PR01', 'PR02', 'PR03', 'PR04', 'PR05', 'PR06', 'PR07', 'PR08',
+             'PR09', 'PR10', 'RT01', 'RT02', 'RT03', 'RT04', 'RT05', 'YD01',
+             'SA01', 'SA02', 'SA03', 'SA04', 'EX01']
+
 warning_codes = ['EX01', 'SA01']
```
Comment (yguclu) [inline ci_tools/process_results.py:19]: Right now `all_codes` contains both error and warning codes, and after specifying `warning_codes` you know that the remaining codes must be errors.

Instead I find it clearer to define two disjoint sets, `error_codes` and `warning_codes`. You can use an assertion to make sure there is no duplicate:

```python
warning_codes = {'EX01', 'SA01'}
error_codes = {'GL01', ..., 'SA03'}

assert warning_codes & error_codes == set()
```
And later in the loop you can write clearly:

```python
if code in warning_codes:
    ...
elif code in error_codes:
    ...
else:
    parsing_errors.append(line)
    # no need for 'continue' statement
```

File: ci_tools/process_results.py [hunk 2]
```diff
@@ -10,25 +10,38 @@
...
+                else:
+                    warnings[file_name].append(msg)
             else:
-                errors[file_name].append(msg)
+                if file_name not in errors:
+                    errors[file_name] = [msg]
+                else:
+                    errors[file_name].append(msg)
+        except ValueError:
+            parsing_errors.append(line)
```
Comment (yguclu) [inline ci_tools/process_results.py:42]: When is this exception raised?

File: ci_tools/list_docs_tovalidate.py [hunk 3]
```diff
@@ -6,55 +6,86 @@
...
+
+        Parameters
+        ----------
+        name : str
+            Absolute name of the inspected object.
+
+        Returns
+        -------
+        bool
+            True if object should be ignored, False otherwise.
```
Comment (yguclu) [inline ci_tools/list_docs_tovalidate.py:21]: Talking about numpydoc... isn't there an excessive indent in this docstring? :sweat_smile:

**Review by yguclu — CHANGES_REQUESTED** [review]
Good job! I have a couple of comments

File: ci_tools/process_results.py (hunk 2 above)
Comment (EmilyBourne) [reply ci_tools/process_results.py:42]: It is raised if `split(':')` fails to find the correct number of colons

File: ci_tools/list_docs_tovalidate.py (hunk 3 above)
Comment (EmilyBourne) [reply ci_tools/list_docs_tovalidate.py:21]: numpydoc doesn't complain about that. But it does complain about a lack of an extended summary

File: ci_tools/process_results.py (hunk 2 above)
Comment (yguclu) [reply ci_tools/process_results.py:42]: Like in this case?
```
>>> a, b, c = (1, 2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: not enough values to unpack (expected 3, got 2)
```
A comment line would be useful.

File: ci_tools/process_results.py (hunk 1 above)
Comment (jalalium) [reply ci_tools/process_results.py:19]: got it! i will fix it!

File: ci_tools/process_results.py (hunk 2 above)
Comment (jalalium) [reply ci_tools/process_results.py:42]: Noted!

**Review by yguclu — APPROVED** [review]

**Pull request merged**